### PR TITLE
fix Issue 13537 - Unions may break immutability

### PR DESCRIPTION
**Pull request by WalterBright**
Reboot of https://github.com/dlang/dmd/pull/5467 which is incomplete.

I'm not satisfied with this yet, it's a WIP.

### CODE REVIEW

**Comment by WalterBright** [comment]
Ready to go when blockers are fixed.

File: src/expression.d [hunk 1]
```diff
@@ -82,74 +82,76 @@ void emplaceExp(T : UnionExp)(T* p, Expression e)
...
     {
-        if (v.overlapped && !sc.intypeof && sc.func && v.type.hasPointers())
+        if (sc.intypeof)
+            return false;
+
+        if (!sc.func)
+            return false;
+
+        if (!sc.func.isSafeBypassingInference())
+            return false;
```
Comment (andralex) [inline src/expression.d]: No  value added by doing these separately.

```d
if (sc.intypeof || !sc.func || !sc.func.isSafeBypassingInference)
    return false;
```

File: src/expression.d [hunk 2]
```diff
@@ -82,74 +82,76 @@ void emplaceExp(T : UnionExp)(T* p, Expression e)
...
+
+        auto ad = v.toParent2().isAggregateDeclaration();
+        if (!ad)
+            return false;
+
+        if (v.overlapped && v.type.hasPointers() && sc.func.setUnsafe())
```
Comment (andralex) [inline src/expression.d]: heh, just like here :)

**Comment by andralex** [comment]
Auto-merge toggled on

**Comment by quickfur** [comment]
Nice!

**Pull request merged**